File: sql_test.go
package dopdb

import (
	"reflect"
	"strings"
	"testing"
)

// The SQL layer is a front end that compiles to the same (filter, sort, page)
// shape FIND uses, so these tests check the COMPILATION — no server needed. The
// resulting filters are then run through the query engine to prove the two agree.

func mustParse(t *testing.T, s string) *SQLQuery {
	t.Helper()
	q, err := ParseSQL(s)
	if err != nil {
		t.Fatalf("ParseSQL(%q): %v", s, err)
	}
	return q
}

func TestSQLCompilesToFilters(t *testing.T) {
	cases := []struct {
		sql  string
		want M
	}{
		{"SELECT * FROM t WHERE a = 1", M{"a": 1.0}},
		{"SELECT * FROM t WHERE a = 'x'", M{"a": "x"}},
		{"SELECT * FROM t WHERE a != 1", M{"a": M{"$ne": 1.0}}},
		{"SELECT * FROM t WHERE a <> 1", M{"a": M{"$ne": 1.0}}},
		{"SELECT * FROM t WHERE a > 1", M{"a": M{"$gt": 1.0}}},
		{"SELECT * FROM t WHERE a >= 1", M{"a": M{"$gte": 1.0}}},
		{"SELECT * FROM t WHERE a < 1", M{"a": M{"$lt": 1.0}}},
		{"SELECT * FROM t WHERE a <= 1", M{"a": M{"$lte": 1.0}}},
		{"SELECT * FROM t WHERE a = TRUE", M{"a": true}},
		{"SELECT * FROM t WHERE a = -3", M{"a": -3.0}},
		{"SELECT * FROM t WHERE a IN (1, 2)", M{"a": M{"$in": []any{1.0, 2.0}}}},
		{"SELECT * FROM t WHERE a NOT IN ('x')", M{"a": M{"$nin": []any{"x"}}}},
		{"SELECT * FROM t WHERE a BETWEEN 1 AND 5", M{"a": M{"$gte": 1.0, "$lte": 5.0}}},
		{"SELECT * FROM t WHERE a IS NULL", M{"a": nil}},
		{"SELECT * FROM t WHERE a IS NOT NULL", M{"a": M{"$ne": nil}}},
		{"SELECT * FROM t WHERE a LIKE 'ab%'", M{"a": M{"$regex": "^ab.*$"}}},
		{"SELECT * FROM t WHERE a ILIKE 'ab%'", M{"a": M{"$regex": "^ab.*$", "$options": "i"}}},
		{"SELECT * FROM t WHERE addr.city = 'NY'", M{"addr.city": "NY"}},
		{"SELECT * FROM t WHERE _id = 'k1'", M{"_id": "k1"}},
	}
	for _, c := range cases {
		q := mustParse(t, c.sql)
		if !reflect.DeepEqual(q.Filter, c.want) {
			t.Errorf("%s\n got %#v\nwant %#v", c.sql, q.Filter, c.want)
		}
	}
}

func TestSQLLogicalStructure(t *testing.T) {
	q := mustParse(t, "SELECT * FROM t WHERE a = 1 AND b = 2")
	and, ok := q.Filter["$and"].([]any)
	if !ok || len(and) != 2 {
		t.Fatalf("AND => %#v", q.Filter)
	}
	q = mustParse(t, "SELECT * FROM t WHERE a = 1 OR b = 2")
	if or, ok := q.Filter["$or"].([]any); !ok || len(or) != 2 {
		t.Fatalf("OR => %#v", q.Filter)
	}
	q = mustParse(t, "SELECT * FROM t WHERE NOT a = 1")
	if nor, ok := q.Filter["$nor"].([]any); !ok || len(nor) != 1 {
		t.Fatalf("NOT => %#v", q.Filter)
	}
	// parentheses bind tighter than the surrounding OR
	q = mustParse(t, "SELECT * FROM t WHERE (a = 1 OR a = 2) AND b = 3")
	if _, ok := q.Filter["$and"]; !ok {
		t.Fatalf("parenthesised => %#v", q.Filter)
	}
}

// Two predicates on the SAME column must both survive — key-merging would
// silently drop one and widen the result.
func TestSQLSameColumnTwiceKeepsBoth(t *testing.T) {
	q := mustParse(t, "SELECT * FROM t WHERE a > 1 AND a < 5")
	d3 := map[string]any{"a": int64(3)}
	d9 := map[string]any{"a": int64(9)}
	if !matchFilter(d3, q.Filter) {
		t.Error("a=3 should match 1 < a < 5")
	}
	if matchFilter(d9, q.Filter) {
		t.Error("a=9 must not match 1 < a < 5")
	}
}

func TestSQLShaping(t *testing.T) {
	q := mustParse(t, "SELECT name, age FROM t ORDER BY age DESC, name ASC LIMIT 10 OFFSET 5")
	if q.Table != "t" {
		t.Errorf("table=%q", q.Table)
	}
	if len(q.Opt.Projection) != 2 || q.Opt.Projection["name"] != 1 {
		t.Errorf("projection=%#v", q.Opt.Projection)
	}
	want := []SortKey{{Field: "age", Asc: false}, {Field: "name", Asc: true}}
	if !reflect.DeepEqual(q.Opt.SortKeys, want) {
		t.Errorf("sort=%#v", q.Opt.SortKeys)
	}
	if q.Opt.Limit != 10 || q.Opt.Skip != 5 {
		t.Errorf("limit=%d skip=%d", q.Opt.Limit, q.Opt.Skip)
	}

	c := mustParse(t, "SELECT COUNT(*) FROM t WHERE a = 1")
	if !c.Count {
		t.Error("COUNT(*) not recognised")
	}
}

// The compiled filter must behave identically to the equivalent hand-written
// one — SQL is a front end, not a second dialect.
func TestSQLAgreesWithTheQueryEngine(t *testing.T) {
	rows := []map[string]any{
		{"_id": "1", "name": "ada", "age": int64(30), "tags": []any{"x"}},
		{"_id": "2", "name": "bob", "age": int64(40)},
		{"_id": "3", "name": "cid", "age": int64(50), "tags": []any{"y"}},
	}
	check := func(sql string, wantIDs ...string) {
		t.Helper()
		q := mustParse(t, sql)
		var got []string
		for _, r := range rows {
			if matchFilter(r, q.Filter) {
				got = append(got, r["_id"].(string))
			}
		}
		if len(got) != len(wantIDs) {
			t.Errorf("%s => %v want %v", sql, got, wantIDs)
			return
		}
		for i := range got {
			if got[i] != wantIDs[i] {
				t.Errorf("%s => %v want %v", sql, got, wantIDs)
				return
			}
		}
	}
	check("SELECT * FROM t WHERE age >= 40", "2", "3")
	check("SELECT * FROM t WHERE age > 30 AND age < 50", "2")
	check("SELECT * FROM t WHERE name IN ('ada','cid')", "1", "3")
	check("SELECT * FROM t WHERE name LIKE 'a%'", "1")
	check("SELECT * FROM t WHERE name ILIKE 'A%'", "1")
	check("SELECT * FROM t WHERE tags IS NULL", "2")
	check("SELECT * FROM t WHERE tags IS NOT NULL", "1", "3")
	check("SELECT * FROM t WHERE NOT age = 40", "1", "3")
	check("SELECT * FROM t WHERE _id = '2'", "2")
	check("SELECT * FROM t WHERE name = 'nobody'")
}

// Everything dopdb deliberately does not implement must be REJECTED with a clear
// message, never silently ignored — a quietly dropped JOIN or GROUP BY would
// return a plausible-looking wrong answer.
func TestSQLRejectsWhatItDoesNotSupport(t *testing.T) {
	cases := []struct {
		sql  string
		want string
	}{
		{"INSERT INTO t VALUES (1)", "read-only"},
		{"UPDATE t SET a = 1", "read-only"},
		{"DELETE FROM t WHERE a = 1", "read-only"},
		{"DROP TABLE t", "DDL"},
		{"SELECT * FROM t JOIN u ON t.a = u.a", "JOIN"},
		{"SELECT * FROM t, u", "multiple tables"},
		{"SELECT * FROM t GROUP BY a", "GROUP BY"},
		{"SELECT SUM(a) FROM t", "COUNT(*)"},
		{"SELECT a AS b FROM t", "aliases"},
		{"SELECT * FROM t WHERE a = NULL", "IS NULL"},
		{"SELECT * FROM t; SELECT * FROM t", "one statement"},
		{"SELECT * FROM t WHERE a = 1 extra", "unexpected"},
		{"SELECT * FROM t LIMIT -1", "non-negative"},
		{"SELECT * FROM `t$evil`", "illegal identifier"},
		{"SELECT * FROM t WHERE a = \"x\"", "literal value"}, // double quotes are identifiers, not strings
	}
	for _, c := range cases {
		_, err := ParseSQL(c.sql)
		if err == nil {
			t.Errorf("%s: expected an error", c.sql)
			continue
		}
		if !strings.Contains(err.Error(), c.want) {
			t.Errorf("%s: error %q does not mention %q", c.sql, err, c.want)
		}
	}
}

// The FROM clause is a security boundary: each collection is authorized
// separately, so a statement must not name a different one.
func TestSQLTableMustMatchCollection(t *testing.T) {
	q := mustParse(t, "SELECT * FROM secrets WHERE a = 1")
	if err := q.CheckTable("notes"); err == nil {
		t.Fatal("FROM secrets must be rejected when serving collection notes")
	}
	if err := q.CheckTable("secrets"); err != nil {
		t.Fatalf("matching table rejected: %v", err)
	}
}

// The compiled filter still has to pass the operator allowlist — SQL is not a
// way around the sanitizer.
func TestSQLOutputPassesSanitizer(t *testing.T) {
	for _, sql := range []string{
		"SELECT * FROM t WHERE a = 1 AND (b > 2 OR c LIKE 'x%')",
		"SELECT * FROM t WHERE a IN (1,2,3) AND d IS NOT NULL",
		"SELECT * FROM t WHERE NOT a BETWEEN 1 AND 2",
	} {
		q := mustParse(t, sql)
		if _, err := SanitizeFilter(q.Filter); err != nil {
			t.Errorf("%s: sanitizer rejected the compiled filter: %v", sql, err)
		}
	}
}

func TestSQLStringLiteralEscaping(t *testing.T) {
	q := mustParse(t, "SELECT * FROM t WHERE name = 'O''Brien'")
	if q.Filter["name"] != "O'Brien" {
		t.Errorf("escaped quote => %#v", q.Filter["name"])
	}
}

func TestLikeToRegexEscapesLiterals(t *testing.T) {
	// a '.' in a LIKE pattern is literal, not a wildcard
	q := mustParse(t, "SELECT * FROM t WHERE a LIKE 'a.c'")
	if matchFilter(map[string]any{"a": "axc"}, q.Filter) {
		t.Error("'.' in LIKE must be literal")
	}
	if !matchFilter(map[string]any{"a": "a.c"}, q.Filter) {
		t.Error("'a.c' should match LIKE 'a.c'")
	}
}

File: ts/test/sql.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";

import { parseSql, checkTable, likeToRegex, SqlError } from "../src/sql.js";
import { matchFilter, type Doc } from "../src/query.js";
import { sanitizeFilter } from "../src/sanitize.js";

// The SQL layer is a front end that compiles to the same (filter, sort, page)
// shape `find` uses, so these tests check the COMPILATION — no server needed.
// They mirror Go's sql_test.go case for case; a divergence between the two
// engines shows up here first.

test("comparisons and predicates compile to the expected filters", () => {
  const cases: [string, unknown][] = [
    ["SELECT * FROM t WHERE a = 1", { a: 1 }],
    ["SELECT * FROM t WHERE a = 'x'", { a: "x" }],
    ["SELECT * FROM t WHERE a != 1", { a: { $ne: 1 } }],
    ["SELECT * FROM t WHERE a <> 1", { a: { $ne: 1 } }],
    ["SELECT * FROM t WHERE a > 1", { a: { $gt: 1 } }],
    ["SELECT * FROM t WHERE a >= 1", { a: { $gte: 1 } }],
    ["SELECT * FROM t WHERE a < 1", { a: { $lt: 1 } }],
    ["SELECT * FROM t WHERE a <= 1", { a: { $lte: 1 } }],
    ["SELECT * FROM t WHERE a = TRUE", { a: true }],
    ["SELECT * FROM t WHERE a = -3", { a: -3 }],
    ["SELECT * FROM t WHERE a IN (1, 2)", { a: { $in: [1, 2] } }],
    ["SELECT * FROM t WHERE a NOT IN ('x')", { a: { $nin: ["x"] } }],
    ["SELECT * FROM t WHERE a BETWEEN 1 AND 5", { a: { $gte: 1, $lte: 5 } }],
    ["SELECT * FROM t WHERE a IS NULL", { a: null }],
    ["SELECT * FROM t WHERE a IS NOT NULL", { a: { $ne: null } }],
    ["SELECT * FROM t WHERE a LIKE 'ab%'", { a: { $regex: "^ab.*$" } }],
    ["SELECT * FROM t WHERE a ILIKE 'ab%'", { a: { $regex: "^ab.*$", $options: "i" } }],
    ["SELECT * FROM t WHERE addr.city = 'NY'", { "addr.city": "NY" }],
    ["SELECT * FROM t WHERE _id = 'k1'", { _id: "k1" }],
  ];
  for (const [sql, want] of cases) {
    assert.deepEqual(parseSql(sql).filter, want, sql);
  }
});

test("AND / OR / NOT / parentheses", () => {
  assert.equal((parseSql("SELECT * FROM t WHERE a = 1 AND b = 2").filter as any).$and.length, 2);
  assert.equal((parseSql("SELECT * FROM t WHERE a = 1 OR b = 2").filter as any).$or.length, 2);
  assert.equal((parseSql("SELECT * FROM t WHERE NOT a = 1").filter as any).$nor.length, 1);
  assert.ok((parseSql("SELECT * FROM t WHERE (a = 1 OR a = 2) AND b = 3").filter as any).$and);
});

// Two predicates on the SAME column must both survive — key-merging would
// silently drop one and widen the result.
test("two predicates on one column both apply", () => {
  const q = parseSql("SELECT * FROM t WHERE a > 1 AND a < 5");
  assert.equal(matchFilter({ a: 3 }, q.filter), true);
  assert.equal(matchFilter({ a: 9 }, q.filter), false);
});

test("projection, ORDER BY, LIMIT, OFFSET and COUNT(*)", () => {
  const q = parseSql("SELECT name, age FROM t ORDER BY age DESC, name ASC LIMIT 10 OFFSET 5");
  assert.equal(q.table, "t");
  assert.deepEqual(q.projection, { name: 1, age: 1 });
  assert.deepEqual(q.sortKeys, [{ field: "age", asc: false }, { field: "name", asc: true }]);
  assert.equal(q.limit, 10);
  assert.equal(q.offset, 5);

  assert.equal(parseSql("SELECT COUNT(*) FROM t WHERE a = 1").count, true);
});

// SQL is a front end, not a second dialect: the compiled filter must behave
// exactly like the equivalent hand-written one.
test("the compiled filter agrees with the query engine", () => {
  const rows: Doc[] = [
    { _id: "1", name: "ada", age: 30, tags: ["x"] },
    { _id: "2", name: "bob", age: 40 },
    { _id: "3", name: "cid", age: 50, tags: ["y"] },
  ];
  const ids = (sql: string) => rows.filter((r) => matchFilter(r, parseSql(sql).filter)).map((r) => r._id);

  assert.deepEqual(ids("SELECT * FROM t WHERE age >= 40"), ["2", "3"]);
  assert.deepEqual(ids("SELECT * FROM t WHERE age > 30 AND age < 50"), ["2"]);
  assert.deepEqual(ids("SELECT * FROM t WHERE name IN ('ada','cid')"), ["1", "3"]);
  assert.deepEqual(ids("SELECT * FROM t WHERE name LIKE 'a%'"), ["1"]);
  assert.deepEqual(ids("SELECT * FROM t WHERE name ILIKE 'A%'"), ["1"]);
  assert.deepEqual(ids("SELECT * FROM t WHERE tags IS NULL"), ["2"]);
  assert.deepEqual(ids("SELECT * FROM t WHERE tags IS NOT NULL"), ["1", "3"]);
  assert.deepEqual(ids("SELECT * FROM t WHERE NOT age = 40"), ["1", "3"]);
  assert.deepEqual(ids("SELECT * FROM t WHERE _id = '2'"), ["2"]);
  assert.deepEqual(ids("SELECT * FROM t WHERE name = 'nobody'"), []);
});

// Everything dopdb deliberately does not implement must be REJECTED with a clear
// message, never silently ignored — a quietly dropped JOIN or GROUP BY would
// return a plausible-looking wrong answer.
test("unsupported SQL is rejected, not ignored", () => {
  const cases: [string, RegExp][] = [
    ["INSERT INTO t VALUES (1)", /read-only/],
    ["UPDATE t SET a = 1", /read-only/],
    ["DELETE FROM t WHERE a = 1", /read-only/],
    ["DROP TABLE t", /DDL/],
    ["SELECT * FROM t JOIN u ON t.a = u.a", /JOIN/],
    ["SELECT * FROM t, u", /multiple tables/],
    ["SELECT * FROM t GROUP BY a", /GROUP BY/],
    ["SELECT SUM(a) FROM t", /COUNT\(\*\)/],
    ["SELECT a AS b FROM t", /aliases/],
    ["SELECT * FROM t WHERE a = NULL", /IS NULL/],
    ["SELECT * FROM t; SELECT * FROM t", /one statement/],
    ["SELECT * FROM t WHERE a = 1 extra", /unexpected/],
    ["SELECT * FROM t LIMIT -1", /non-negative/],
    ["SELECT * FROM `t$evil`", /illegal identifier/],
    ['SELECT * FROM t WHERE a = "x"', /literal value/], // double quotes are identifiers
  ];
  for (const [sql, want] of cases) {
    assert.throws(() => parseSql(sql), want, sql);
  }
});

// A rejected statement is the caller's mistake: it must surface as 400, not 500.
test("SqlError carries 400 / validation", () => {
  try {
    parseSql("NOPE");
    assert.fail("should have thrown");
  } catch (e) {
    assert.ok(e instanceof SqlError);
    assert.equal((e as SqlError).status, 400);
    assert.equal((e as SqlError).code, "validation");
  }
});

// The FROM clause is a security boundary: each collection is authorized
// separately, so a statement must not name a different one.
test("FROM must match the collection being served", () => {
  const q = parseSql("SELECT * FROM secrets WHERE a = 1");
  assert.throws(() => checkTable(q, "notes"), /does not match/);
  assert.doesNotThrow(() => checkTable(q, "secrets"));
});

// SQL is not a way around the operator allowlist.
test("compiled filters pass the sanitizer", () => {
  for (const sql of [
    "SELECT * FROM t WHERE a = 1 AND (b > 2 OR c LIKE 'x%')",
    "SELECT * FROM t WHERE a IN (1,2,3) AND d IS NOT NULL",
    "SELECT * FROM t WHERE NOT a BETWEEN 1 AND 2",
  ]) {
    assert.doesNotThrow(() => sanitizeFilter(parseSql(sql).filter));
  }
});

test("'' escapes a quote inside a string literal", () => {
  assert.equal((parseSql("SELECT * FROM t WHERE name = 'O''Brien'").filter as any).name, "O'Brien");
});

test("LIKE metacharacters are escaped, wildcards are not", () => {
  assert.equal(likeToRegex("a.c"), "^a\\.c$");
  assert.equal(likeToRegex("a%c_d"), "^a.*c.d$");
  const q = parseSql("SELECT * FROM t WHERE a LIKE 'a.c'");
  assert.equal(matchFilter({ a: "axc" }, q.filter), false);
  assert.equal(matchFilter({ a: "a.c" }, q.filter), true);
});
